### ISSUE TITLE
Multiple entries per article in list view

### ISSUE THREAD

**Issue by davidskalinder** [davidskalinder/mpeds-coder#34]
@matthewsmorgan reports: When I coded testing articles, saved, then confirmed things were saved by navigating to the List from the UI Home, I'm seeing multiple entries for individual articles. This was not occurring last time I tested the UI (you can see previously one entry per article from articles last updated 1/24 and older).

@davidskalinder replies: Yeah, that's weird.  I'll investigate further.

**Comment by davidskalinder** [comment]
So @matthewsmorgan, somehow all the articles about Wisconsin got assigned to you (and to Chaeyoon, John, and @olderwoman) twice, which I'm about 99% sure is why they're coming up twice in your list.

I assigned those articles to all the testers the first time, and I think they probably got assigned the second time when @olderwoman was assigning the articles to the trainees from the admin interface (because the non-black Wisconsin articles, which I deleted afterward, are not assigned to you anymore).

I can't decide whether this should be treated as a bug or not though--it seems like maybe we'll want to assign articles to coders twice sometimes (if they screwed it up the first time, for example).  @olderwoman, can you clarify what we want this to do?

**Comment by olderwoman** [comment]
Yes, that is exactly what happened. I re-assigned them because I did the assignment in one big batch as that was easiest for me. My fault.

**Comment by davidskalinder** [comment]
Hmm, this is also causing some very strange behavior with the way articles are marked as done.  In particular, when duplicate entries are in a queue and (at least when) one of the duplicates is already marked as done but the other isn't, the article can't be marked as done unless new information is added; if all information is removed, then the timestamp for the first instance only of the duplicate is article is updated, and the coder can't move on, until some information is positively added to the article.

This is very strange and almost certainly buggy behavior, although it shouldn't bother us unless/until we start having articles appear more than once in a queue; so we can probably live without fixing it this week, at least . . .

**Comment by olderwoman** [comment]
Well the cause of the behavior is my laziness, but I guess it does reveal some sort of underlying bug. Although probably not an especially important bug, given what had to happen for it to occur.

**Comment by davidskalinder** [comment]
Okay, @olderwoman and @alexhanna, I think I need your input on this one . . .:

After more thinking and testing (but not having read through the code yet), I'm still not sure I completely understand the way this issue works.  However, I'm starting to think that _maybe_ this is the intended functionality?

As far as I can tell, the way it works now is:

- Admins can assign an article to a queue as many times as they want
- When coders see an assigned article the first time, they can skip it
- When coders see an assigned article after the first time, they must change it to a non-blank entry (even if they do it by changing a field and then changing it back to the way it was)

So maybe that's what we want?  If we eventually want to force coders to return to articles they've seen before and take action on them?

So I won't change anything until I hear back from y'all.  The whole way queue management is set up now seems a little awkward, but in the short term we can manage it however we want in the back end (though this will require caution to not delete records of finished articles).  So if we want to change the way this issue works then maybe it'll be part of a larger set of changes to the queue management that we'll make down the road?

**Comment by davidskalinder** [comment]
> Well the cause of the behavior is my laziness, but I guess it does reveal some sort of underlying bug. Although probably not an especially important bug, given what had to happen for it to occur. 

If it is actually a bug, we should fix it at some point.  It's pretty easy to accidentally hit that button twice in a row, and at the moment there's no way to remove assigned articles in the interface (which I'm about to add as another issue).  So we can avoid it by being careful and fixing it in the back end, but neither of those are reliable long-term solutions...

**Comment by davidskalinder** [comment]
Another thought about this one: if we do in fact want to allow multiple entries of single articles per coder, we should at least enumerate the entries in the database (and not just with the table key: it should be something like rep_no or some such).

**Comment by olderwoman** [comment]
Looks like I have not followed the discussion on this point.

**Comment by davidskalinder** [comment]
From today's meeting:

- Is the current behavior desirable?
  - Maybe, but maybe not worth the complexity
  - Get the extra timestamp from the duplicates
    - But we get this from the coded fields

In conclusion, this is almost certainly a bug that nobody wants, so we should put some constraints on the database column to forbid duplicates. (@alexhanna, I'm guessing you agree?)

**Comment by alexhanna** [comment]
Sorry for the late remarks on this. The queue table should indeed have a unique constraint on coder_id and article_id. I never intended for coders to ever have the same article in their queue more than once.

**Comment by davidskalinder** [comment]
Gotcha, thanks.  I almost convinced myself that it might be intended functionality, but then everybody else talked me out of it. :)  I'll build this on a branch as close to upstream as possible.

**Comment by davidskalinder** [comment]
So @alexhanna, I just noticed this:

https://github.com/davidskalinder/mpeds-coder/blob/89e6f6223e219481006cdbd3520e75781db72cb0/assign_lib.py#L58

The fact that there's a `distinct` in there makes me nervous...  Are you sure that duplicates were never allowed in that table?  Even for like the old pass 1 / pass 2 code?  If not, and if we always expected no duplicates in the table, then why put the distinct in there?

Sorry if I'm being paranoid, but I don't want to make a change that breaks something unexpected...

**Comment by davidskalinder** [comment]
Whoops, I'm an idiot: that query is for all users, so the distinct just removes duplicates across users, which of course are expected.  Sorry for the false alarm.

**Comment by davidskalinder** [comment]
Um.:

https://github.com/davidskalinder/mpeds-coder/blob/89e6f6223e219481006cdbd3520e75781db72cb0/models.py#L155

This makes me think that something must be going wrong with how sqlalchemy is putting in the constraints, and makes this a bigger issue than just adding the constraints in.  For a temporary fix, I think I should try running ALTER TABLE to make sure the data doesn't get out of control...

**Comment by davidskalinder** [comment]
Ran `ALTER TABLE event_creator_queue ADD CONSTRAINT unique1 UNIQUE (article_id, coder_id);` in dev deployment.  Adding article IDs that are already in a user's queue now generates a (annoyingly nonspecfic) 500 error.  So, yeah, looks good enough for now anyway.

**Comment by davidskalinder** [comment]
Annd ran the same thing in production DB.  Didn't run it in testing because I know there's dupes in those queues, so it should choke.
